polkadot/node/core/backing/src/lib.rs

```diff
@@ -981,6 +996,75 @@ async fn handle_active_leaves_update<Context>(
 	Ok(())
 }
 
+macro_rules! try_runtime_api {
+	($x: expr) => {
+		match $x {
+			Ok(x) => x,
+			Err(err) => {
+				// Only bubble up fatal errors.
+				error::log_error(Err(Into::<runtime::Error>::into(err).into()))?;
+
+				// We can't do candidate validation work if we don't have the
+				// requisite runtime API data. But these errors should not take
+				// down the node.
+				return Ok(None)
+			},
+		}
+	};
+}
+
+fn core_index_from_statement(
+	validator_groups: &[Vec<ValidatorIndex>],
+	group_rotation_info: &GroupRotationInfo,
+	cores: &[CoreState],
+	statement: &SignedFullStatementWithPVD,
+) -> Option<CoreIndex> {
+	let compact_statement = statement.as_unchecked();
+	let candidate_hash = CandidateHash(*compact_statement.unchecked_payload().candidate_hash());
+
+	let n_cores = cores.len();
+
+	gum::trace!(target: LOG_TARGET, ?group_rotation_info, ?statement, ?validator_groups, n_cores = ?cores.len() , ?candidate_hash, "Extracting core index from statement");
+
+	let statement_validator_index = statement.validator_index();
+	for (group_index, group) in validator_groups.iter().enumerate() {
+		for validator_index in group {
+			if *validator_index == statement_validator_index {
+				// First check if the statement para id matches the core assignment.
+				let core_index =
+					group_rotation_info.core_for_group(GroupIndex(group_index as u32), n_cores);
+
+				if core_index.0 as usize > n_cores {
+					gum::warn!(target: LOG_TARGET, ?candidate_hash, ?core_index, n_cores, "Invalid CoreIndex");
+					return None
+				}
+
+				if let StatementWithPVD::Seconded(candidate, _pvd) = statement.payload() {
+					let candidate_para_id = candidate.descriptor.para_id;
+					let assigned_para_id = match &cores[core_index.0 as usize] {
+						CoreState::Free => {
+							gum::debug!(target: LOG_TARGET, ?candidate_hash, "Invalid CoreIndex, core is not assigned to any para_id");
+							return None
+						},
+						CoreState::Occupied(occupied) => occupied.candidate_descriptor.para_id,
```

Comment:
This looks wrong. We should be checking [next_up_on_available](https://github.com/paritytech/polkadot-sdk/blob/fbe1ad516de65b20807c13b502051103547ad538/polkadot/primitives/src/v6/mod.rs#L1020). (Worth checking the collator protocol, maybe it is wrong there as well.)

Reply:
You're right. I am discovering more and more places where async backing (and by extension elastic scaling) has not been optimised for on-demand.
See https://github.com/paritytech/polkadot-sdk/issues/3141, https://github.com/paritytech/polkadot-sdk/issues/3327, https://github.com/paritytech/polkadot-sdk/blob/d7862aa8c9b4f8be1d4330bc11c742bf48d407f6/polkadot/node/network/collator-protocol/src/validator_side/mod.rs#L515 and https://github.com/paritytech/polkadot-sdk/blob/d7862aa8c9b4f8be1d4330bc11c742bf48d407f6/polkadot/node/core/backing/src/lib.rs#L1065

All of these places (may be more) would work fine as long as the on-demand parachains maintain their core affinities. Whenever there's a switch, cores will be left unoccupied for a relay parent.

I think we should optimise this as a separate issue

Reply:
Yes although do it right from the start in new changes.

Reply:
I wonder whether we should also take into account next_up_on_timeout.
It adds quite a lot of complexity. I think we're better off keeping things simple and accepting than on rare timeout occasions, we'll see an empty core for a relay chain block duration

Reply:
done